### PR TITLE
Issue-184/integrate-golanglint-only

### PR DESCRIPTION
**Pull request by AugustasV**
#### Summary
#184 Only integrated Golangci-lint. I am not able to fix those linter errors, it's just too much of them ( deadcode, errcheck, golint, gosec and others) only fixed some whitespaces

#### Ticket Link

  Integrate Golangci-lint https://github.com/mattermost/mattermost-plugin-mscalendar/issues/184

### CODE REVIEW

**Comment by jasonblais** [comment]
@hanzei @jfrerich what would be a good way to proceed?

**Comment by hanzei** [comment]
Talked with @AugustasV offline. He will disable the failing linters for now.

**Comment by AugustasV** [comment]
@hanzei sure right now build is successful.

**Comment by hanzei** [comment]
I did enable most of the linter and fixed the errors. @mickmister @jfrerich Please take a look when you have the time. It's a lot to reivew.

File: server/api/post_action.go [hunk 1]
```diff
@@ -96,15 +96,15 @@ func (api *api) postActionRespond(w http.ResponseWriter, req *http.Request) {
...
 	}
 
-	p, err := api.PluginAPI.GetPost(postID)
-	if err != nil {
-		utils.SlackAttachmentError(w, "Error: Failed to update the post: "+err.Error())
+	p, appErr := api.PluginAPI.GetPost(postID)
+	if appErr != nil {
+		utils.SlackAttachmentError(w, "Error: Failed to update the post: "+appErr.Error())
 		return
 	}
```
Comment (hanzei) [inline server/api/post_action.go:103]: Please double check that the logic is correct now. With `err` L113 makes little sense.

Comment (AugustasV) [reply server/api/post_action.go:103]: @hanzei that looks fine for me

File: server/command/unsubscribe.go [hunk 2]
```diff
@@ -17,5 +13,6 @@ func (c *Command) unsubscribe(parameters ...string) (string, bool, error) {
 	if err != nil {
 		return "", false, err
 	}
-	return fmt.Sprintf("You have unsubscribed from events."), false, nil
```
Comment (jfrerich) [inline server/command/unsubscribe.go:20]: I love how these are caught with linters :)

**Review by jfrerich — APPROVED** [review]
Thanks @AugustasV! 

@hanzei thanks for all the lint fixes also.  That was a lot of changes, but all made sense.  I love the linters <3

**Comment by jasonblais** [comment]
@mickmister kind reminder to help with a review :)

File: server/command/command.go [hunk 3]
```diff
@@ -7,10 +7,9 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/pkg/errors"
-
 	"github.com/mattermost/mattermost-server/v5/model"
 	"github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/pkg/errors"
```
Comment (mickmister) [inline server/command/command.go:12]: This seems weird. We usually have the non-Mattermost in groups above the Mattermost packages. WIll the linter allow you to order/group dependencies like this?

```go
import (
    "strings"

    "github.com/pkg/errors"	

    "github.com/mattermost/mattermost-server/v5/model"
    "github.com/mattermost/mattermost-server/v5/plugin"

    "github.com/mattermost/mattermost-plugin-mscalendar/server/config"
)
```

File: server/mscalendar/welcome_flow.go [hunk 4]
```diff
@@ -6,15 +6,15 @@ import (
 	"github.com/mattermost/mattermost-plugin-mscalendar/server/utils/flow"
 )
 
-type welcomeFlow struct {
+type WelcomeFlow struct {
```
Comment (mickmister) [inline server/mscalendar/welcome_flow.go:9]: Does this need to become public?

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @AugustasV! I have some questions about the `Makefile`, and a request to group/order package imports if they are from a mattermost repo.

File: server/command/command.go (hunk 3 above)
Comment (mickmister) [reply server/command/command.go:12]: This should be applied to all Go files that are applicable

File: Makefile [hunk 5]
```diff
@@ -1,27 +1,13 @@
...
-BUILD_DATE = $(shell date -u)
-BUILD_HASH = $(shell git rev-parse HEAD)
-BUILD_HASH_SHORT = $(shell git rev-parse --short HEAD)
-LDFLAGS += -X "main.BuildDate=$(BUILD_DATE)"
-LDFLAGS += -X "main.BuildHash=$(BUILD_HASH)"
-LDFLAGS += -X "main.BuildHashShort=$(BUILD_HASH_SHORT)"
-GO_TEST_FLAGS ?= -race
-GO_BUILD_FLAGS ?=
-GOBUILD = $(GO) build $(GOFLAGS) -ldflags '$(LDFLAGS)'
-GOTEST = $(GO) test $(GOFLAGS) -ldflags '$(LDFLAGS)'
```
[the remark this answers is not in the document]
Comment (AugustasV) [reply Makefile:24]: -ldflags exist in old build

File: build/custom.mk [hunk 6]
```diff
@@ -1,6 +1 @@
-# Include custom targets and environment variables here
-ifndef MM_RUDDER_WRITE_KEY
-MM_RUDDER_WRITE_KEY = 1d5bMvdrfWClLxgK1FvV3s4U1tg
-endif
-
-LDFLAGS += -X "github.com/mattermost/mattermost-plugin-mscalendar/server/utils/telemetry.rudderWriteKey=$(MM_RUDDER_WRITE_KEY)"
```
[the remark this answers is not in the document]
Comment (AugustasV) [reply build/custom.mk:6]: ok reverted

File: Makefile [hunk 7]
```diff
@@ -44,90 +30,34 @@ all: check-style test dist
...
+## Run golangci-lint on codebase.
+.PHONY: golangci-lint
+golangci-lint:
+	@if ! [ -x "$$(command -v golangci-lint)" ]; then \
+		echo "golangci-lint is not installed. Please see https://github.com/golangci/golangci-lint#install for installation instructions."; \
+		exit 1; \
+	fi; \
 
-## Generates mock golang interfaces for testing
-mock:
```
[the remark this answers is not in the document]
Comment (AugustasV) [reply Makefile:97]: ok

File: Makefile [hunk 8]
```diff
@@ -238,77 +157,17 @@ endif
...
-.PHONY: sd
-sd: server-debug
-
-# server-debug builds and deploys a debug version of the plugin for your architecture.
-# Then resets the plugin to pick up the changes.
-.PHONY: server-debug
-server-debug: server-debug-deploy reset
-
-.PHONY: server-debug-deploy
-server-debug-deploy: validate-go-version
```
[the remark this answers is not in the document]
Comment (AugustasV) [reply Makefile:264]: reverted

File: server/mscalendar/welcome_flow.go (hunk 4 above)
Comment (AugustasV) [reply server/mscalendar/welcome_flow.go:9]: I don't know what is this, in vs code it shows no change 
image

File: server/command/command.go (hunk 3 above)
Comment (hanzei) [reply server/command/command.go:12]: > We usually have the non-Mattermost in groups above the Mattermost packages. 

is that really the convention? I would argue that treat `github.com/mattermost/mattermost-server/v5` in the same way as other library. Also, we don't enforce any rules around this making it hard to check.

File: server/mscalendar/welcome_flow.go (hunk 4 above)
Comment (hanzei) [reply server/mscalendar/welcome_flow.go:9]: Because `NewWelcomeFlow` returns a `WelcomeFlow`. Go lint insists of exporting structs that are returned.

**Review by mickmister — APPROVED** [review]
LGTM thanks @AugustasV!

**Review by jfrerich — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann Not sure what to test for this PR. It mostly contains build changes.

**Comment by DHaussermann** [comment]
@AugustasV I don't see any lint errors here but I do see what looks like a change in behavior here. if I just type `make` I see 
```
➜  mattermost-plugin-mscalendar git:(AugustasV-master) ✗ make
go install github.com/golang/mock/mockgen
mockgen -destination server/jobs/mock_cluster/mock_cluster.go github.com/mattermost/mattermost-plugin-api/cluster JobPluginAPI
mockgen -destination server/mscalendar/mock_mscalendar/mock_mscalendar.go github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar MSCalendar
mockgen -destination server/mscalendar/mock_welcomer/mock_welcomer.go -package mock_welcomer github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar Welcomer
mockgen -destination server/mscalendar/mock_plugin_api/mock_plugin_api.go -package mock_plugin_api github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar PluginAPI
mockgen -destination server/remote/mock_remote/mock_remote.go github.com/mattermost/mattermost-plugin-mscalendar/server/remote Remote
mockgen -destination server/remote/mock_remote/mock_client.go github.com/mattermost/mattermost-plugin-mscalendar/server/remote Client
mockgen -destination server/utils/bot/mock_bot/mock_poster.go github.com/mattermost/mattermost-plugin-mscalendar/server/utils/bot Poster
mockgen -destination server/utils/bot/mock_bot/mock_admin.go github.com/mattermost/mattermost-plugin-mscalendar/server/utils/bot Admin
mockgen -destination server/utils/bot/mock_bot/mock_logger.go github.com/mattermost/mattermost-plugin-mscalendar/server/utils/bot Logger
mockgen -destination server/store/mock_store/mock_store.go github.com/mattermost/mattermost-plugin-mscalendar/server/store Store
```
and then it stops it does not go on to run unit test and build the binary. There are no issues when I type `make all` any thoughts on what could cause this change or if it's an issue?

cc @hanzei

**Comment by hanzei** [comment]
@DHaussermann For plugin that have a custom makefile at `build/custom.mk` `make` will always run the first command in the custom makefile. This is the expected behaviour in this case. The command you are looking for is `make all`.

**Comment by DHaussermann** [comment]
Thanks @hanzei I figured it may not be an issue that you have to use `make all` and not just `make` But it felt odd to me that on `master` branch this behavior is different. Does this also make sense to you?

**Comment by hanzei** [comment]
Yes. This PR was the one that introduced `build/custom.mk`.

**Review by DHaussermann — APPROVED** [review]
Tested and passed

No issues will build or deployment. 
LGTM!

Thanks @AugustasV for adding this! Much appreciated.

**Pull request merged**